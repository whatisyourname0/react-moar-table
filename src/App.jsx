import { useMemo, useState } from 'react';
import {
  useBlockLayout,
  useResizeColumns,
  useSortBy,
  useTable,
} from 'react-table';
import { useSticky } from 'react-table-sticky';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import useScrollInfo from './hooks/useScrollInfo';
import textTruncate from './styles/textTruncate';
import customScrollbar from './styles/scrollbar';

function App({ columns, data, minWidth, width, hoveredColor, clickedColor }) {
  const [hoveredCell, setHoveredCell] = useState(undefined);
  const [clickedCell, setClickedCell] = useState(undefined);
  const [scrollInfo, setRef] = useScrollInfo();
  const defaultColumn = useMemo(
    () => ({
      minWidth: Number(minWidth),
      width: Number(width),
    }),
    [minWidth, width],
  );

  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    useTable(
      { columns, data, defaultColumn },
      useSticky,
      useBlockLayout,
      useResizeColumns,
      useSortBy,
    );

  return (
    <TableWrapper
      hoveredCell={hoveredCell}
      hoveredColor={hoveredColor}
      clickedCell={clickedCell}
      clickedColor={clickedColor}
      onMouseLeave={() => setHoveredCell(undefined)}
      ref={setRef}
      xRatio={scrollInfo.x.percentage}
    >
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map((headerGroup, idx) => (
            <tr key={idx} {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map((column, idx) => (
                <th
                  key={idx}
                  {...column.getHeaderProps(column.getSortByToggleProps())}
                  onClick={() => {
                    setClickedCell(idx + 1);
                    column.toggleSortBy(!column.isSortedDesc);
                  }}
                  onMouseOver={() => setHoveredCell(idx + 1)}
                >
                  {column.render('Header')}
                  <SortedSpan>
                    {column.isSorted
                      ? column.isSortedDesc
                        ? '\t⬇︎'
                        : '\t⬆︎'
                      : ''}
                  </SortedSpan>
                  {column.canResize && (
                    <Resizer
                      {...column.getResizerProps()}
                      onClick={(event) => event.stopPropagation()}
                    />
                  )}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map((row, idx) => {
            prepareRow(row);
            return (
              <tr key={idx} {...row.getRowProps()}>
                {row.cells.map((cell, idx) => (
                  <td
                    key={idx}
                    onMouseOver={() => setHoveredCell(idx + 1)}
                    {...cell.getCellProps()}
                  >
                    <TextWrapper>{cell.render('Cell')}</TextWrapper>
                  </td>
                ))}
              </tr>
            );
          })}
        </tbody>
      </table>
    </TableWrapper>
  );
}

App.propTypes = {
  columns: PropTypes.array.isRequired,
  data: PropTypes.array.isRequired,
  minWidth: PropTypes.number || PropTypes.string,
  width: PropTypes.number || PropTypes.string,
  hoveredColor: PropTypes.string,
  clickedColor: PropTypes.string,
};

App.defaultProps = {
  minWidth: 80,
  width: 150,
  hoveredColor: '#f1f1f1',
  clickedColor: '#e6f6ff',
};

export default App;

const TextWrapper = styled.div`
  ${textTruncate};
`;

const Resizer = styled.div`
  position: absolute;
  top: 0;
  right: 0;

  width: 0;
  height: 100%;

  background: #bdbdbd;
  z-index: 1;
  touch-action: none;

  transition: 0.2s ease;
`;

const SortedSpan = styled.span`
  font-size: 16px;
`;

const addTableEdge = css`
  box-shadow: 3px 3px 3px #939393;
`;

// Style table at here!
// If wrapping inner jsx element with styled, react throws warning!
const TableWrapper = styled.div`
  margin: 16px;
  font-size: 16px;

  overflow-x: auto;
  white-space: nowrap;
  user-select: none;
  ${customScrollbar};

  table {
    border-spacing: 0;
    text-align: center;
    will-change: background-color;

    [data-sticky-td] {
      position: sticky !important;
      background-color: #fafafa;
    }

    [data-sticky-last-left-td] {
      ${(props) => (props.xRatio !== 0 ? addTableEdge : null)};
    }

    thead {
      overflow-y: auto;
      overflow-x: hidden;

      th {
        border-bottom: 1px solid black;
        background-color: #f4f4f4;

        :hover {
          ${Resizer} {
            width: 7.5px;
          }
        }

        :active {
          ${Resizer} {
            width: 15px;
            background-color: #8e8e8e;
          }
        }
      }
    }

    tbody {
      overflow-y: scroll;
      overflow-x: hidden;
    }

    th,
    td {
      position: relative;
      margin: 0;
      padding: 10px 15px;
      overflow: hidden;
      background: #f9f9f9;

      transition: 0.2s ease;
    }

    th:nth-child(${(props) => props.hoveredCell}),
    td:nth-child(${(props) => props.hoveredCell}),
    tbody tr:hover {
      background-color: ${(props) => props.hoveredColor};

      td {
        background-color: ${(props) => props.hoveredColor};
      }
    }

    th:nth-child(${(props) => props.clickedCell}),
    td:nth-child(${(props) => props.clickedCell}) {
      background-color: ${(props) => props.clickedColor} !important;
    }
  }
`;
