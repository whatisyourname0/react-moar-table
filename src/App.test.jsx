import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import App from './App';

afterEach(cleanup);

const testColumn = [
  {
    Header: 'Run ID',
    accessor: 'info.run_id',
  },
];

const testData = [
  {
    info: {
      run_id: 'test1',
    },
  },
  {
    info: {
      run_id: 'test2',
    },
  },
];

test('Render Test', () => {
  const utils = render(<App columns={testColumn} data={testData} />);
  expect(utils.container).toMatchSnapshot();
});

test('Inner Text Test', () => {
  render(<App columns={testColumn} data={testData} />);

  const columnID = screen.getByText('Run ID');
  const runID1 = screen.getByText('test1');
  const runID2 = screen.getByText('test2');

  expect(columnID).toBeInTheDocument();
  expect(runID1).toBeInTheDocument();
  expect(runID2).toBeInTheDocument();
});
