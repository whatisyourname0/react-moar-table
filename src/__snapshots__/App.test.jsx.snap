// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Test 1`] = `
.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c3 {
  position: absolute;
  top: 0;
  right: 0;
  width: 0;
  height: 100%;
  background: #bdbdbd;
  z-index: 1;
  touch-action: none;
  -webkit-transition: 0.2s ease;
  transition: 0.2s ease;
}

.c1 {
  font-size: 16px;
}

.c0 {
  margin: 16px;
  font-size: 16px;
  overflow-x: auto;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0::-webkit-scrollbar {
  width: 6px;
  height: 6px;
}

.c0::-webkit-scrollbar-track {
  background-color: transparent;
}

.c0::-webkit-scrollbar-thumb {
  border-radius: 3px;
  background-color: rgb(180,180,180);
}

.c0::-webkit-scrollbar-thumb:hover {
  background-color: rgb(150,150,150);
}

.c0::-webkit-scrollbar-thumb:active {
  background-color: rgb(120,120,120);
}

.c0::-webkit-scrollbar-button {
  width: 0;
  height: 0;
}

.c0 table {
  border-spacing: 0;
  text-align: center;
  will-change: background-color;
}

.c0 table [data-sticky-td] {
  position: -webkit-sticky !important;
  position: sticky !important;
  background-color: #fafafa;
}

.c0 table [data-sticky-last-left-td] {
  box-shadow: 3px 3px 3px #939393;
}

.c0 table thead {
  overflow-y: auto;
  overflow-x: hidden;
}

.c0 table thead th {
  border-bottom: 1px solid black;
  background-color: #f4f4f4;
}

.c0 table thead th:hover .c2 {
  width: 7.5px;
}

.c0 table thead th:active .c2 {
  width: 15px;
  background-color: #8e8e8e;
}

.c0 table tbody {
  overflow-y: scroll;
  overflow-x: hidden;
}

.c0 table th,
.c0 table td {
  position: relative;
  margin: 0;
  padding: 10px 15px;
  overflow: hidden;
  background: #f9f9f9;
  -webkit-transition: 0.2s ease;
  transition: 0.2s ease;
}

.c0 table th:nth-child(),.c0 table td:nth-child(),
.c0 table tbody tr:hover {
  background-color: #f1f1f1;
}

.c0 table th:nth-child() td,.c0 table td:nth-child() td,
.c0 table tbody tr:hover td {
  background-color: #f1f1f1;
}

.c0 table th:nth-child(),.c0 table td:nth-child() {
  background-color: #e6f6ff !important;
}

<div>
  <div
    class="c0"
  >
    <table
      role="table"
    >
      <thead>
        <tr
          role="row"
          style="display: flex; width: 150px;"
        >
          <th
            colspan="1"
            role="columnheader"
            style="display: inline-block; box-sizing: border-box; width: 150px; position: relative; cursor: pointer;"
            title="Toggle SortBy"
          >
            Run ID
            <span
              class="c1"
            />
            <div
              class="c2 c3"
              draggable="false"
              role="separator"
              style="cursor: col-resize;"
            />
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
          style="display: flex; width: 150px;"
        >
          <td
            role="cell"
            style="display: inline-block; box-sizing: border-box; width: 150px;"
          >
            <div
              class="c4"
            >
              test1
            </div>
          </td>
        </tr>
        <tr
          role="row"
          style="display: flex; width: 150px;"
        >
          <td
            role="cell"
            style="display: inline-block; box-sizing: border-box; width: 150px;"
          >
            <div
              class="c4"
            >
              test2
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Resizer Test 1`] = `
.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c3 {
  position: absolute;
  top: 0;
  right: 0;
  width: 0;
  height: 100%;
  background: #bdbdbd;
  z-index: 1;
  touch-action: none;
  -webkit-transition: 0.2s ease;
  transition: 0.2s ease;
}

.c1 {
  font-size: 16px;
}

.c0 {
  margin: 16px;
  font-size: 16px;
  overflow-x: auto;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0::-webkit-scrollbar {
  width: 6px;
  height: 6px;
}

.c0::-webkit-scrollbar-track {
  background-color: transparent;
}

.c0::-webkit-scrollbar-thumb {
  border-radius: 3px;
  background-color: rgb(180,180,180);
}

.c0::-webkit-scrollbar-thumb:hover {
  background-color: rgb(150,150,150);
}

.c0::-webkit-scrollbar-thumb:active {
  background-color: rgb(120,120,120);
}

.c0::-webkit-scrollbar-button {
  width: 0;
  height: 0;
}

.c0 table {
  border-spacing: 0;
  text-align: center;
  will-change: background-color;
}

.c0 table [data-sticky-td] {
  position: -webkit-sticky !important;
  position: sticky !important;
  background-color: #fafafa;
}

.c0 table [data-sticky-last-left-td] {
  box-shadow: 3px 3px 3px #939393;
}

.c0 table thead {
  overflow-y: auto;
  overflow-x: hidden;
}

.c0 table thead th {
  border-bottom: 1px solid black;
  background-color: #f4f4f4;
}

.c0 table thead th:hover .c2 {
  width: 7.5px;
}

.c0 table thead th:active .c2 {
  width: 15px;
  background-color: #8e8e8e;
}

.c0 table tbody {
  overflow-y: scroll;
  overflow-x: hidden;
}

.c0 table th,
.c0 table td {
  position: relative;
  margin: 0;
  padding: 10px 15px;
  overflow: hidden;
  background: #f9f9f9;
  -webkit-transition: 0.2s ease;
  transition: 0.2s ease;
}

.c0 table th:nth-child(),.c0 table td:nth-child(),
.c0 table tbody tr:hover {
  background-color: #f1f1f1;
}

.c0 table th:nth-child() td,.c0 table td:nth-child() td,
.c0 table tbody tr:hover td {
  background-color: #f1f1f1;
}

.c0 table th:nth-child(),.c0 table td:nth-child() {
  background-color: #e6f6ff !important;
}

<div
  className="c0"
  onMouseLeave={[Function]}
>
  <table
    role="table"
  >
    <thead>
      <tr
        role="row"
        style={
          Object {
            "display": "flex",
            "width": "150px",
          }
        }
      >
        <th
          colSpan={1}
          onClick={[Function]}
          onMouseOver={[Function]}
          role="columnheader"
          style={
            Object {
              "boxSizing": "border-box",
              "cursor": "pointer",
              "display": "inline-block",
              "position": "relative",
              "width": "150px",
            }
          }
          title="Toggle SortBy"
        >
          Run ID
          <span
            className="c1"
          />
          <div
            className="c2 c3"
            draggable={false}
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            role="separator"
            style={
              Object {
                "cursor": "col-resize",
              }
            }
          />
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
        style={
          Object {
            "display": "flex",
            "width": "150px",
          }
        }
      >
        <td
          onMouseOver={[Function]}
          role="cell"
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "width": "150px",
            }
          }
        >
          <div
            className="c4"
          >
            test1
          </div>
        </td>
      </tr>
      <tr
        role="row"
        style={
          Object {
            "display": "flex",
            "width": "150px",
          }
        }
      >
        <td
          onMouseOver={[Function]}
          role="cell"
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "width": "150px",
            }
          }
        >
          <div
            className="c4"
          >
            test2
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
